### PR TITLE
MULE-10260: Add endpoint for retrieving keys by config or resolver

### PR DESCRIPTION
**Pull request by pablomusumeci**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 1]
```diff
@@ -50,13 +52,19 @@
   private static final String EXCEPTION_RESOLVING_COMPONENT_METADATA =
       "An exception occurred while resolving metadata for component '%s'";
   private static final String COMPONENT_NOT_METADATA_AWARE = "Component is not MetadataAware, no information available";
+  private static final String CANNOT_FETCH_METADATA_KEYS = "Cannot fetch MetadataKeys from component [%s]";
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: `"Component [%s ]is not MetadataKeysAware, no information available"` ?

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 2]
```diff
@@ -50,13 +52,19 @@
...
   private static final String SOURCE_NOT_FOUND = "Flow doesn't contain a message source";
   private static final String PROCESSOR_NOT_FOUND = "Processor doesn't exist in the given index [%s]";
+  private static final String CONFIG_NOT_FOUND = "Configuration named [%s] doesn't exist";
+
 
   @Inject
   private MuleContext muleContext;
 
+  @Inject
+  protected ConnectionManagerAdapter connectionManager;
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: why? this looks like old code

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 3]
```diff
@@ -100,6 +108,17 @@ public void initialise() throws InitialisationException {
     return exceptionHandledMetadataFetch(componentId, MetadataAware::getMetadataKeys, EXCEPTION_RESOLVING_METADATA_KEYS);
   }
 
+  @Override
+  public MetadataResult<Map<String, Set<MetadataKey>>> getMetadataKeys(ConfigurationId configurationId) {
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: Can't we use a ComponentId? I know it has `getFlowName` in its api, but at some point we could migrate to using only the `getComponentPath`. Once we implement #MULE-9496 we can come up with a convention over how a path for a global element looks like vs something inside a flow.

If we don't change this now, at least we should add a `TODO MULE-9496`

Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: `MetadataResult<Map<String, Set<MetadataKey>>>` may require a custom serializer, we can check this when implementing the agent service, but we should keep it in mind.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+import org.apache.commons.collections.map.HashedMap;
+
+/**
+ * Adds the capability to expose all the {@link MetadataKey}s associated with the {@link StaticConfigurationProvider}'s
+ * components.
+ *
+ * @since 1.0
+ */
+public final class DefaultConfigurationProviderMetadataAdapter<T> extends StaticConfigurationProvider<T>
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java]: Linked to the comments on mule-api I think this should be a `ConfigurationProviderMetadataAdapter` which implements `MetadataKeysAware`

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 5]
```diff
@@ -175,6 +194,20 @@ private MetadataAware findMetadataAwareComponent(ComponentId componentId) throws
     }
   }
 
+  private ConfigurationProviderMetadataAdapter findMetadataKeyAwareComponent(String configName)
+      throws InvalidComponentIdException {
+    try {
+      ConfigurationProviderMetadataAdapter<Object> configurationProvider = muleContext.getRegistry().lookupObject(configName);
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: `ConfigurationProviderMetadataAdapter` => `MetadataKeysAware`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+  @Inject
+  private MuleMetadataManager metadataManager;
+
+  @Inject
+  protected ConnectionManagerAdapter connectionManager;
+
+  public MetadataResult<Map<String, Set<MetadataKey>>> getMetadataKeys() throws MetadataResolvingException {
+
+    Map<String, Set<MetadataKey>> keys = new HashedMap();
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java]: `HashedMap` => `HashMap`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import org.mule.runtime.extension.api.introspection.metadata.NullMetadataResolver;
+import org.mule.runtime.extension.api.runtime.ConfigurationInstance;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.apache.commons.collections.map.HashedMap;
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java]: nopes

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+public final class DefaultConfigurationProviderMetadataAdapter<T> extends StaticConfigurationProvider<T>
+    implements ConfigurationProviderMetadataAdapter<T> {
+
+  public DefaultConfigurationProviderMetadataAdapter(String name, RuntimeConfigurationModel model,
+                                                     ConfigurationInstance<T> configuration) {
+    super(name, model, configuration);
+  }
+
+  @Inject
+  private MuleMetadataManager metadataManager;
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java]: this before constructor

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  private void addComponentKeys(List<? extends ComponentModel> components, MetadataContext metadataContext,
+                                Map<String, Set<MetadataKey>> keys)
+      throws MetadataResolvingException, ConnectionException {
+    for (ComponentModel component : components) {
+      MetadataKeysResolver keysResolver =
+          ((MetadataEnrichableModel) component).getMetadataResolverFactory().getKeyResolver();
+
+      String resolverName = keysResolver.getClass().getName();
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultConfigurationProviderMetadataAdapter.java]: `IntrospectionUtils.getAliasName(keysResolver.getClass())`

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 10]
```diff
@@ -8,11 +8,12 @@
 
 import static java.lang.String.format;
 import static org.mule.runtime.core.config.i18n.MessageFactory.createStaticMessage;
-
 import org.mule.runtime.api.metadata.ComponentId;
+import org.mule.runtime.api.metadata.MetadataKeysAware;
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: rename package to datasense... Change all the occurences of Metadata to DataSense

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 11]
```diff
@@ -50,9 +50,13 @@
   private static final String EXCEPTION_RESOLVING_COMPONENT_METADATA =
       "An exception occurred while resolving metadata for component '%s'";
   private static final String COMPONENT_NOT_METADATA_AWARE = "Component is not MetadataAware, no information available";
+  private static final String COMPONENT_NOT_METADATA_KEY_AWARE =
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: consider the naming changes in the mule-api review

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 12]
```diff
@@ -96,10 +100,21 @@ public void initialise() throws InitialisationException {
...
+  public MetadataResult<MetadataKeysContainer> getMetadataKeys(ComponentId componentId) {
     return exceptionHandledMetadataFetch(componentId, MetadataAware::getMetadataKeys, EXCEPTION_RESOLVING_METADATA_KEYS);
   }
 
+  @Override
+  public MetadataResult<MetadataKeysContainer> getMetadataKeysForConfig(ComponentId configurationId) {
+    try {
+      return findMetadataKeyAwareComponent(configurationId.getComponentPath()).getMetadataKeys();
+    } catch (InvalidComponentIdException e) {
+      return MetadataResult.failure(e);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: static import

File: core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java [hunk 13]
```diff
@@ -175,6 +190,20 @@ private MetadataAware findMetadataAwareComponent(ComponentId componentId) throws
     }
   }
 
+  private MetadataKeysAware findMetadataKeyAwareComponent(String configName)
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/metadata/MuleMetadataManager.java]: rename method

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ConfigurationProviderMetadataAdapter.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+import javax.inject.Inject;
+
+/**
+ * Adds the capability to expose all the {@link MetadataKey}s associated with the {@link StaticConfigurationProvider}'s
+ * components.
+ *
+ * @since 1.0
+ */
+public final class ConfigurationProviderMetadataAdapter<T> extends StaticConfigurationProvider<T>
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ConfigurationProviderMetadataAdapter.java]: Metadata -> DataSense

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ConfigurationProviderMetadataAdapter.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+                                              ConfigurationInstance<T> configuration) {
+    super(name, model, configuration);
+  }
+
+  public MetadataResult<MetadataKeysContainer> getMetadataKeys() throws MetadataResolvingException {
+
+    MetadataKeysContainerBuilder keysBuilder = new MetadataKeysContainerBuilder();
+    MetadataContext metadataContext = getMetadataContext();
+    try {
+      addComponentKeys(this.getModel().getOperationModels(), metadataContext, keysBuilder);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ConfigurationProviderMetadataAdapter.java]: redundant use of this.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ConfigurationProviderMetadataAdapter.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+          ((MetadataEnrichableModel) component).getMetadataResolverFactory().getKeyResolver();
+
+      String resolverName = getAliasName(keysResolver.getClass());
+      if (!(keysResolver instanceof NullMetadataResolver) && !keysBuilder.containsResolver(resolverName)) {
+        keysBuilder.add(resolverName, keysResolver.getMetadataKeys(metadataContext));
+      }
+    }
+  }
+
+
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/ConfigurationProviderMetadataAdapter.java]: empty line

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MetadataTypeUtils.java [hunk 17]
```diff
@@ -133,4 +136,8 @@ public static boolean isExtensible(MetadataType metadataType) {
   public static boolean isEnum(MetadataType metadataType) {
     return metadataType.getAnnotation(EnumAnnotation.class).isPresent();
   }
+
+  public static Set<MetadataKey> getKeysFromContainer(MetadataKeysContainer metadataKeysContainer) {
+    return metadataKeysContainer.getKeys(metadataKeysContainer.getResolvers().iterator().next()).get();
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MetadataTypeUtils.java]: is no clear from the signature what this method does. I'd spectrum it to squash the keys from all the resolvrrs into a single set, not to take only the first. if this is for testing only then can add it to the basetestcase as util method

Comment (pablomusumeci) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MetadataTypeUtils.java]: That was my first approach (adding it to MetadataExtensionFunctionalTestCase) but then I needed the same method in another test from another module (in extension-support and extension-spring-support).

**Comment by marianogonzalez** [comment]
jenkins validate

**Pull request merged**